src/test/assembly/asm/s390x-types.rs

```diff
@@ -0,0 +1,126 @@
+// min-llvm-version: 10.0.1
+// revisions: s390x
+// assembly-output: emit-asm
+//[s390x] compile-flags: --target s390x-unknown-linux-gnu
+//[s390x] needs-llvm-components: systemz
+
+#![feature(no_core, lang_items, rustc_attrs, repr_simd)]
+#![crate_type = "rlib"]
+#![no_core]
+#![allow(asm_sub_register, non_camel_case_types)]
+
+#[rustc_builtin_macro]
+macro_rules! asm {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! concat {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! stringify {
+    () => {};
+}
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "copy"]
+trait Copy {}
+
+type ptr = *const i32;
+
+impl Copy for i8 {}
+impl Copy for u8 {}
+impl Copy for i16 {}
+impl Copy for i32 {}
+impl Copy for i64 {}
+impl Copy for f32 {}
+impl Copy for f64 {}
+impl Copy for ptr {}
+
+extern "C" {
+    fn extern_func();
+    static extern_static: u8;
+}
+
+// Hack to avoid function merging
+extern "Rust" {
+    fn dont_merge(s: &str);
+}
+
+macro_rules! check { ($func:ident, $ty:ty, $class:ident, $mov:literal) => {
+    #[no_mangle]
+    pub unsafe fn $func(x: $ty) -> $ty {
+        dont_merge(stringify!(func));
+
+        let y;
+        asm!(concat!($mov," {}, {}"), out($class) y, in($class) x);
+        y
+    }
+};}
+
+macro_rules! check_reg { ($func:ident, $ty:ty, $reg:tt, $mov:literal) => {
+    #[no_mangle]
+    pub unsafe fn $func(x: $ty) -> $ty {
+        dont_merge(stringify!(func));
+
+        let y;
+        asm!(concat!($mov, " ", $reg, ", ", $reg), lateout($reg) y, in($reg) x);
+        y
+    }
+};}
+
+// CHECK-LABEL: sym_fn_32:
+// CHECK: #APP
+// CHECK: brasl %r14, extern_func
+// CHECK: #NO_APP
+#[cfg(s390x)]
+#[no_mangle]
+pub unsafe fn sym_fn_32() {
+    asm!("brasl %r14, {}", sym extern_func);
+}
+
+// CHECK-LABEL: sym_static:
+// CHECK: #APP
+// CHECK: brasl %r14, extern_static
+// CHECK: #NO_APP
+#[no_mangle]
+pub unsafe fn sym_static() {
+    asm!("brasl %r14, {}", sym extern_static);
+}
+
+// CHECK-LABEL: reg_i8:
+// CHECK: #APP
+// CHECK: lgr %r{{[0-9]+}}, %r{{[0-9]+}}
+// CHECK: #NO_APP
+check!(reg_i8, i8, reg, "lgr");
+
+// CHECK-LABEL: reg_i16:
+// CHECK: #APP
+// CHECK: lgr %r{{[0-9]+}}, %r{{[0-9]+}}
+// CHECK: #NO_APP
+check!(reg_i16, i16, reg, "lgr");
+
+// CHECK-LABEL: reg_i32:
+// CHECK: #APP
+// CHECK: lgr %r{{[0-9]+}}, %r{{[0-9]+}}
+// CHECK: #NO_APP
+check!(reg_i32, i32, reg, "lgr");
+
+// CHECK-LABEL: reg_i64:
+// CHECK: #APP
+// CHECK: lgr %r{{[0-9]+}}, %r{{[0-9]+}}
+// CHECK: #NO_APP
+check!(reg_i64, i64, reg, "lgr");
```

Comment:
Add a test with pointer types. Refer to the other tests for examples.

Reply:
This is the test I have for the pointer type, similar to `reg_ptr` in `x86-types.rs` :

```
// CHECK-LABEL: reg_ptr:
// CHECK: #APP
// CHECK: lgr %r{{[0-9]+}}, %r{{[0-9]+}}
// CHECK: #NO_APP
check!(reg_ptr, ptr, reg, "lgr");
```

and `stderr`:

```
/home/linux1/rust/src/test/assembly/asm/s390x-types.rs:128:17: error: CHECK-LABEL: expected string not found in input
// CHECK-LABEL: reg_ptr:
                ^
/home/linux1/rust/build/s390x-unknown-linux-gnu/test/assembly/asm/s390x-types.s390x/s390x-types.s:174:9: note: scanning from here
reg_f64:
        ^
/home/linux1/rust/build/s390x-unknown-linux-gnu/test/assembly/asm/s390x-types.s390x/s390x-types.s:204:9: note: possible intended match here
 .globl reg_ptr
```

I am not sure why `reg_ptr` is `.globl`, nor how to go about resolving this issue.

Reply:
Can you paste the full contents of `/home/linux1/rust/build/s390x-unknown-linux-gnu/test/assembly/asm/s390x-types.s390x/s390x-types.s`?

Reply:
```
	.text
	.file	"s390x_types.63e18542-cgu.0"
	.section	.text.sym_fn_32,"ax",@progbits
	.globl	sym_fn_32
	.p2align	4
	.type	sym_fn_32,@function
sym_fn_32:
	.cfi_startproc
	stmg	%r14, %r15, 112(%r15)
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -160
	.cfi_def_cfa_offset 320
	#APP
	brasl	%r14, extern_func
	#NO_APP
	lmg	%r14, %r15, 272(%r15)
	br	%r14
.Lfunc_end0:
	.size	sym_fn_32, .Lfunc_end0-sym_fn_32
	.cfi_endproc

	.section	.text.sym_static,"ax",@progbits
	.globl	sym_static
	.p2align	4
	.type	sym_static,@function
sym_static:
	.cfi_startproc
	stmg	%r14, %r15, 112(%r15)
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -160
	.cfi_def_cfa_offset 320
	#APP
	brasl	%r14, extern_static
	#NO_APP
	lmg	%r14, %r15, 272(%r15)
	br	%r14
.Lfunc_end1:
	.size	sym_static, .Lfunc_end1-sym_static
	.cfi_endproc

	.section	.text.reg_i8,"ax",@progbits
	.globl	reg_i8
	.p2align	4
	.type	reg_i8,@function
reg_i8:
	.cfi_startproc
	stmg	%r13, %r15, 104(%r15)
	.cfi_offset %r13, -56
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -160
	.cfi_def_cfa_offset 320
	lr	%r13, %r2
	larl	%r2, .L__unnamed_1
	lghi	%r3, 4
	brasl	%r14, dont_merge@PLT
	#APP
	lgr	%r2, %r13
	#NO_APP
	lmg	%r13, %r15, 264(%r15)
	br	%r14
.Lfunc_end2:
	.size	reg_i8, .Lfunc_end2-reg_i8
	.cfi_endproc

	.section	.text.reg_i16,"ax",@progbits
	.globl	reg_i16
	.p2align	4
	.type	reg_i16,@function
reg_i16:
	.cfi_startproc
	stmg	%r13, %r15, 104(%r15)
	.cfi_offset %r13, -56
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -160
	.cfi_def_cfa_offset 320
	lr	%r13, %r2
	larl	%r2, .L__unnamed_1
	lghi	%r3, 4
	brasl	%r14, dont_merge@PLT
	#APP
	lgr	%r2, %r13
	#NO_APP
	lmg	%r13, %r15, 264(%r15)
	br	%r14
.Lfunc_end3:
	.size	reg_i16, .Lfunc_end3-reg_i16
	.cfi_endproc

	.section	.text.reg_i32,"ax",@progbits
	.globl	reg_i32
	.p2align	4
	.type	reg_i32,@function
reg_i32:
	.cfi_startproc
	stmg	%r13, %r15, 104(%r15)
	.cfi_offset %r13, -56
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -160
	.cfi_def_cfa_offset 320
	lr	%r13, %r2
	larl	%r2, .L__unnamed_1
	lghi	%r3, 4
	brasl	%r14, dont_merge@PLT
	#APP
	lgr	%r2, %r13
	#NO_APP
	lmg	%r13, %r15, 264(%r15)
	br	%r14
.Lfunc_end4:
	.size	reg_i32, .Lfunc_end4-reg_i32
	.cfi_endproc

	.section	.text.reg_i64,"ax",@progbits
	.globl	reg_i64
	.p2align	4
	.type	reg_i64,@function
reg_i64:
	.cfi_startproc
	stmg	%r13, %r15, 104(%r15)
	.cfi_offset %r13, -56
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -160
	.cfi_def_cfa_offset 320
	lgr	%r13, %r2
	larl	%r2, .L__unnamed_1
	lghi	%r3, 4
	brasl	%r14, dont_merge@PLT
	#APP
	lgr	%r2, %r13
	#NO_APP
	lmg	%r13, %r15, 264(%r15)
	br	%r14
.Lfunc_end5:
	.size	reg_i64, .Lfunc_end5-reg_i64
	.cfi_endproc

	.section	.text.reg_f32,"ax",@progbits
	.globl	reg_f32
	.p2align	4
	.type	reg_f32,@function
reg_f32:
	.cfi_startproc
	stmg	%r14, %r15, 112(%r15)
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -168
	.cfi_def_cfa_offset 328
	std	%f8, 160(%r15)
	.cfi_offset %f8, -168
	ler	%f8, %f0
	larl	%r2, .L__unnamed_1
	lghi	%r3, 4
	brasl	%r14, dont_merge@PLT
	#APP
	ler	%f0, %f8
	#NO_APP
	ld	%f8, 160(%r15)
	lmg	%r14, %r15, 280(%r15)
	br	%r14
.Lfunc_end6:
	.size	reg_f32, .Lfunc_end6-reg_f32
	.cfi_endproc

	.section	.text.reg_f64,"ax",@progbits
	.globl	reg_f64
	.p2align	4
	.type	reg_f64,@function
reg_f64:
	.cfi_startproc
	stmg	%r14, %r15, 112(%r15)
	.cfi_offset %r14, -48
	.cfi_offset %r15, -40
	aghi	%r15, -168
	.cfi_def_cfa_offset 328
	std	%f8, 160(%r15)
	.cfi_offset %f8, -168
	ldr	%f8, %f0
	larl	%r2, .L__unnamed_1
	lghi	%r3, 4
	brasl	%r14, dont_merge@PLT
	#APP
	ldr	%f0, %f8
	#NO_APP
	ld	%f8, 160(%r15)
	lmg	%r14, %r15, 280(%r15)
	br	%r14
.Lfunc_end7:
	.size	reg_f64, .Lfunc_end7-reg_f64
	.cfi_endproc

	.type	.L__unnamed_1,@object
	.section	.rodata.cst4,"aM",@progbits,4
	.p2align	1
.L__unnamed_1:
	.ascii	"func"
	.size	.L__unnamed_1, 4

	.globl	reg_ptr
	.type	reg_ptr,@function
.set reg_ptr, reg_i64
	.section	".note.GNU-stack","",@progbits
```